File: python/read_addresses.py
#!
import pandas as pd
from pandas.tseries.offsets import DateOffset

path = 'C:/Users/edavis67/OneDrive - DXC Production/Documents/' \
    'Industrialized AI Badge/Guild_Project/'

with open(path + 'output.csv', 'r') as op:
    lines = op.readlines()
    keys = [i.strip() for i in lines[0].split(", ")][:6]
    keys = [key for key in keys[:6] if key != 'zip_code']
    fmtLines = []
    errLines = []
    for line in lines[1:]:
        ln = line.split(", ")
        date = ln[0] + ", "+ ln[1]
        addr = ln[3]
        price = ln[4]
        lon = float(ln[5])
        lat = float(ln[6])
        values = [date, addr, price, lon, lat]
        fmtLines.append(dict(zip(keys, values)))

df = pd.DataFrame.from_dict(fmtLines)
df['amount'] = df['amount'].str.replace('[\,, \$]', '').astype('float')
df['address'] = df['address'].str.replace('.', '').str.replace(',', '')

date1 = pd.to_datetime(df['date'],
    format = "%B %d, %Y", exact = False, errors = 'coerce')
date2 = pd.to_datetime(df['date'],
    format = "%b. %d, %Y", exact = False, errors = 'coerce')
df['date'] = pd.Series(date1).fillna(value = date2)
df['L1Month'] = df.date - DateOffset(months = 1)
df['L1Month'] = pd.to_datetime(df['L1Month'],
    format = "%Y-%m-%d", unit = 'D',
    exact = False, errors = 'coerce')
df['L1Year'] = df.L1Month - DateOffset(years = 1)
df['L1Year'] = pd.to_datetime(df['L1Year'],
    format = "%Y-%m-%d", unit = 'D',
    exact = False, errors = 'coerce')

df = df[df.L1Month > '2012-02-01']
df = df[df[['address', 'L1Month', 'L1Year', 'longitude', 'latitude']]
    .notna().all(axis = 'columns')]
df = df.drop_duplicates(subset = ['address', 'date']).sort_values(by = 'date')
df.to_parquet(path + 'property.parquet')

File: R/read_addresses.R





library(tidyverse)
# ------------------------------------------------------------------------------
# There are commas in the fields that confuse the standard read_csv function, so
# we have to clean the data. Read the original csv file line by line and save it
# to an R object---a character vector:
output <- read_lines("./output.csv")
# ------------------------------------------------------------------------------
# We write our own function to digest the data line by line from`output`:
make_df <- function(.) {
  # Break down each line and add a newline so the tokenizer works, and unlist:
  tk <- unlist(tokenize(paste0(., "\n")))
  # Format the date. The tokenizer separates the month and day from the year,
  # and some of the months include period abbreviations, while others have the
  # full name---try both:
  dt <- as.Date(paste0(tk[1], ", ", tk[2]), tryFormats = c("%B %d, %Y", "%b. %d, %Y"))
  # The zip code is always the 3rd, and the address is always the 4th:
  zipcode <- tk[3]
  addr <- tk[4]
  # Most of the amounts have commas to delimit the millions and thousands. If
  # you only grab the 5th and 6th components, you miss anything over a million
  # dollars, and sometimes they are non-zero, so you check that there are three
  # characters:
  if (str_length(trimws(tk[6])) == 3 && str_length(trimws(tk[7])) == 3) {
    price <- paste0(tk[5:7], collapse = "")
    longitude <- tk[8]
    latitude <- tk[9]
    buyer_seller <- tk[10:(length(tk) - 2)]
    # Sometimes the amount does not have a thousands component, so we check the
    # first four characters to see if they are consistent with a longitude:
  } else if (str_sub(tk[6], 1, 4) == "-90." || str_sub(tk[6], 1, 4) == "-89.") {
    price <- paste0(tk[5], collapse = "")
    longitude <- tk[6]
    latitude <- tk[7]
    buyer_seller <- tk[8:(length(tk) - 2)]
    # Everything else should be < $1000000 and > $999:
  } else {
    price <- paste0(tk[5:6], collapse = "")
    longitude <- tk[7]
    latitude <- tk[8]
    buyer_seller <- tk[9:(length(tk) - 2)]
  }
  # The last two components are the lot and district:
  district <- tk[length(tk) - 1]
  lot <- tk[length(tk)]
  # Form a tibble (a dataframe that does not coerce characters into factors or
  # add row names to dataframes). The buyer_seller is a combined list of the
  # buyers and sellers. As they get mangled in the tokenization, they catch
  # everthing that cannot be classified otherwise:
  df <- tibble(date = dt, zip_code = zipcode, address = addr,
                   amount = price, longitude = longitude, latitude = latitude,
                   buyer_seller = list(buyer_seller), district = district, lot = lot, )
  return(df)
}
# Forms the dataframe by concatenating the individual rows formed from the above
# function: (Combined below with the data transformations.) 
# df <- map_df(output[-1], make_df)
# ------------------------------------------------------------------------------
# Three function wrappers to process the columns. They add '[EMPTY]' to the list
# of NA's. Wrapping the function is an alternative to passing the argument list
# in the mutate_at function below:
parse_int <- function(.) parse_integer(., na = c("", "NA", "[EMPTY]"))
parse_num <- function(.) parse_number(., na = c("", "NA", "[EMPTY]"))
parse_chr <- function(.) parse_character(., na = c("", "NA", "[EMPTY]"))
# ------------------------------------------------------------------------------
# Creates the dataframe, passing it to the functions that turn them into
# integer, number, and character vectors:
df <- map_df(output[-1], make_df) %>%
  mutate_at(vars(zip_code, district, lot), parse_int) %>% 
  mutate_at(vars(amount, longitude, latitude), parse_num) %>%
  mutate_at(vars(address), parse_chr)
# ------------------------------------------------------------------------------
tmp <- df
# Drop the buyer and seller to make exporting more convenient:
tmp$buyer_seller <- NULL
# Remove inconvenient punctuation if necessary:
tmp$address <- str_remove_all(tmp$address, "[.,]")
# Summary to check if the data make sense:
summary(tmp)
# Export to rds and csv formats:
write_rds(tmp, "./property.rds")
write_csv(tmp, "./property.csv")



